numpy/lib/format.py

```diff
@@ -586,7 +586,10 @@ def write_array(fp, array, version=None, allow_pickle=True, pickle_kwargs=None):
         The version number of the format. None means use the oldest
         supported version that is able to store the data.  Default: None
     allow_pickle : bool, optional
-        Whether to allow writing pickled data. Default: True
+        Whether to allow writing pickled data. Default: None
+        In numpy <= 1.16.1 the value defaults to True. This default is being
+        deprecated during 1.17, with a value of None now triggering a warning,
+        and in 1.18 allow_pickle=None will default to False.
```

Comment:
What value are we gaining by changing the save behavior?

Reply:
@ivanov do you mind if I/we just push updates to the PR very liberally to move these things along a bit quicker, given that distros would probably like to use it soon?

Reply:
@seberg I think this part of the PR is not clear. There is no danger in **writing** pickled data, the problem is **reading** pickled data which may execute arbitrary code. The [python documentation](https://docs.python.org/3/library/pickle.html) agrees with this assesment

Reply:
Yes, we should undo this part. I thought for a bit that we could possibly add some notification, but frankly there seems no point.

Reply:
What we gain by changing the save behavior is the roundtrip of defaults are retained. I wrote about that in the [first comment of this PR](https://github.com/numpy/numpy/pull/12889#issuecomment-459137063)

> The vulnerability only applies to loading, but I thought it'd be awkward if the default parameters did not round-trip save-load with default parameters.

Reply:
I disagree. No one writes data without the intention of reading it back. I think without this users will only stub their toes on the new default on load and won't know at write time if they are doing something that we now think they should have to opt into explicitly.

Reply:
But we allow people to read it back, they simply need to be aware of what they are doing. For instance, Word macros are considered dangerous. Most operating systems notice when you download a document and will open it read-only. People who create documents are not prevented from saving their work.

Reply:
@ivanov, that is why we could discuss warnings. But overall I think I am also for just leaving it out now. If you read it in again yourself, it hardly matters when you see the warning/error, on writing or on reading (you stub your toes exactly once either way).

The only thing it would help to warn on writing would be if you are not aware that you are saving objects to begin with (it can happen on accident after all). It might also be nice for users who intend to share the file to be warned. But in most cases they will share it anyway because the other side will trust them.

Reply:
I haven't used MS Word for a long time, but it used to just give you a big scary warning when you load a document with macros. Had a warning or an option to remove macros been present at save time, I think it would have been a better user experience: "You are saving a document with macros. You, or anyone you share the document with will get a big scary warning when trying to reopen this in the future." with an explicit opt in to suppress that warning and save those macros to begin with.

Again, we are not preventing saving of peoples work. You [indicated earlier that this would only be the case of some other package fiddling with the global warnings filters](https://github.com/numpy/numpy/pull/12889/files#r258650142). On lines [613-614](https://github.com/numpy/numpy/blob/53ba1220acccc0121cb206d742f6cad4056a2361/numpy/lib/format.py#L614-L615), there is a code path with a UserWarning for storing arrays in format 2.0 being able to be read by NumPy >= 1.9. It sounds like you're saying we should not have warnings in any save-related code path?

Reply:
My gut feeling is that it is important to warn "savers" of problems "loaders" may experience, otherwise there will be no indication that things may go wrong.  Round-tripping is a pretty standard test that your I/O works correctly; does it make much sense to have a `save` function of which the output cannot be loaded by default?  What is the downside to making users explicitly set a flag to  request saving objects?

Reply:
The downside is that a long running script not saving the final result would be very painful. On the other hand, there is not much downside to a UserWarning (possibly after already saving successfully as pointed out below). That would be an in-between choice which is not quite symmetric but puts informative for users that pay attention to warnings.

Reply:
I could live with a UserWarning.

Reply:
UserWarning added for both saving and loading

Reply:
Sorry, I guess this was a bit too much back and forth/discussion. I think we want a FutureWarning on loading, since we would want to make setting it mandatory in the future (since otherwise it is not really safe as the warning is "too late"), but a UserWarning on saving, because saving is not a security issue and the warning would be only there for a bit more symmetry and to inform users.

Reply:
fixed